File: src/game/poly.ts

export class Block {
    public x: number;
    public y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }
}

/**
 * Poly !!!!.
 */
export class Poly {
    blocks: Array<Block>;

    _hashCode: number = 0;

    constructor() {
        this.blocks = new Array<Block>();
        this._hashCode = undefined;
    }

    static fromBlocks(blocks: Block[]): Poly {
        const poly = new Poly();
        poly.blocks = blocks;
        return poly;
    }

    get length(): number {
        return this.blocks.length;
    }

    clonePoly(): Poly {
        // ensure current piece is deeply cloned
        const cloneBlocks = new Array<Block>(this.length);
        for (let i = 0; i < this.length; i++) {
            const block = new Block(this.blocks[i].x, this.blocks[i].y);
            cloneBlocks[i] = block;
        }
        const clone = new Poly();
        clone.blocks = cloneBlocks;
        clone._hashCode = this.getHashCode();
        return clone;
    }

    createArray(): Poly[] {
        const polyArray = new Array<Poly>();
        polyArray.push(this);
        return polyArray;
    }

    rotateClockwise(): Poly {

        const clone = this.clonePoly();

        // rotate about the first block
        for (let i = 1; i < clone.length; i++) {
            const x = clone.blocks[i].x - clone.blocks[0].x;
            const y = clone.blocks[i].y - clone.blocks[0].y;
            clone.blocks[i].x = -y + clone.blocks[0].x;
            clone.blocks[i].y = x + clone.blocks[0].y;
        }
        return clone;
    }

    rotateAntiClockwise(): Poly {

        const clone = this.clonePoly();

        for (let i = 1; i < clone.length; i++) {
            const x = clone.blocks[i].x - clone.blocks[0].x;
            const y = clone.blocks[i].y - clone.blocks[0].y;
            clone.blocks[i].x = y + clone.blocks[0].x;
            clone.blocks[i].y = -x + clone.blocks[0].y;
        }
        return clone;
    }

    /**
     * Creates a hex string colour for the poly. The colour will be
     * consistent as the poly moves and rotates.
     */
    createPolyColor(): string {
        const hashCode = this.getHashCode();
        let color = Math.abs(hashCode);
        const limit = 13777215;
        if (color > limit) {
            color = color % limit;
        }
        let code = color.toString(16);
        while (code.length < 6) {
            code = "0" + code;
        }
        return "#" + code;
    }

    createPreviewPiece(): Poly {
        const previewPiece = this.clonePoly();
        for (let i = 0; i < previewPiece.blocks.length; i++) {
            const block = previewPiece.blocks[i];
            block.x += Math.floor(previewPiece.blocks.length / 2) - 1;
            block.y += Math.floor(previewPiece.blocks.length / 2) - 1;
        }
        return previewPiece;
    }

    getHash(): String {
        const blockHashes = new Array();
        for (let i = 0; i < this.length; i++) {
            const polyHash = JSON.stringify(this.blocks[i]);
            blockHashes.push(polyHash);
        }
        return JSON.stringify(Array.from(blockHashes).sort());
    }

    getHashCode(): number {

        if (!this._hashCode) {

            let hashCode = 0;
            const hashStr = this.getHash();
            if (hashStr.length == 0) return hashCode;
            for (let i = 0; i < hashStr.length; i++) {
                const char = hashStr.charCodeAt(i);
                hashCode = ((hashCode << 5) - hashCode) + char;
                hashCode = hashCode & hashCode; // Convert to 32bit integer
            }
            this._hashCode = hashCode;
        }
        return this._hashCode;
    }
}

File: src/poly.rs
use std::hash::Hash;

use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct Block {
    pub x: i32,
    pub y: i32,

}

impl Block {

    pub fn new(x: i32, y: i32) -> Block {
        Block {
            x,y
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct Poly {
    pub blocks: Vec<Block>,

    _hashCode: i32,

}

impl Poly{

    pub fn from_blocks(blocks: Vec<Block>) -> Poly {
        Poly {
            blocks,
            ..Default::default()
        }
    }

    pub fn length(&self) -> usize {
        self.blocks.len()
    }

    /*
    clonePoly(): Poly {
        // ensure current piece is deeply cloned
        var cloneBlocks = new Array<Block>(this.length);
        for (var i = 0; i < this.length; i++) {
            var block = new Block(this.blocks[i].x, this.blocks[i].y);
            cloneBlocks[i] = block;
        }
        var clone = new Poly();
        clone.blocks = cloneBlocks;
        clone._hashCode = this.getHashCode();
        return clone;
    }

    createArray(): Poly[] {
        var polyArray = new Array<Poly>();
        polyArray.push(this);
        return polyArray;
    }
    */

    pub fn rotateClockwise(&self) -> Poly {

        let mut clone = self.clone();

        // rotate about the first block
        for  i in 1..clone.length() {
            let x = clone.blocks[i].x - clone.blocks[0].x;
            let y = clone.blocks[i].y - clone.blocks[0].y;
            clone.blocks[i].x = -y + clone.blocks[0].x;
            clone.blocks[i].y = x + clone.blocks[0].y;
        }
        clone
    }

    pub fn rotateAntiClockwise(&self) -> Poly {

        let mut clone = self.clone();

        for i in 1..clone.length() {
            let x = clone.blocks[i].x - clone.blocks[0].x;
            let y = clone.blocks[i].y - clone.blocks[0].y;
            clone.blocks[i].x = y + clone.blocks[0].x;
            clone.blocks[i].y = -x + clone.blocks[0].y;
        }
        clone
    }

    /**
     * Creates a hex string colour for a given poly. The colour will be
     * consistent as the poly moves and rotates.
     * @param poly 
     */
    pub fn create_poly_color() -> String {
        "#ffffff".to_string()
    }
    /*
    pub fn createPolyColor() -> string {
        var hashCode = this.getHashCode();
        var color = Math.abs(hashCode);
        var limit = 13777215;
        if (color > limit) {
            color = color % limit;
        }
        var code = color.toString(16);
        while (code.length < 6) {
            code = "0" + code;
        }
        return "#" + code;
    }
    */

    /*
    createPreviewPiece(): Poly {
        var previewPiece = this.clonePoly();
        for (var i = 0; i < previewPiece.blocks.length; i++) {
            var block = previewPiece.blocks[i];
            block.x += Math.floor(previewPiece.blocks.length / 2) - 1;
            block.y += Math.floor(previewPiece.blocks.length / 2) - 1;
        }
        return previewPiece;
    }
    */

    pub fn getHash(&self) -> String {
        let mut block_hashes = Vec::new();
        for block in &self.blocks {
            let block_hash = serde_json::to_string(&block).unwrap();
            block_hashes.push(block_hash);
        }
        block_hashes.sort();
        serde_json::to_string(&block_hashes).unwrap()
    }

    /*
    getHashCode(): number {

        if (this._hashCode === null) {

            var hashCode = 0;
            var hashStr = this.getHash();
            if (hashStr.length == 0) return hashCode;
            for (var i = 0; i < hashStr.length; i++) {
                var char = hashStr.charCodeAt(i);
                hashCode = ((hashCode << 5) - hashCode) + char;
                hashCode = hashCode & hashCode; // Convert to 32bit integer
            }
            this._hashCode = hashCode;
        }
        return this._hashCode;
    }
*/
}

impl PartialEq for Poly {
    fn eq(&self, other: &Self) -> bool {
        self.getHash() == other.getHash()
    }
}
impl Eq for Poly {
}
impl Hash for Poly {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.getHash().hash(state);
    }
}